kv/kvserver: TestStoreRangeSplitAndMergeWithGlobalReads failed

kv/kvserver.TestStoreRangeSplitAndMergeWithGlobalReads [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/14005299?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/14005299?buildTab=artifacts#/) on master @ [cc6ca026319024800395293b0fb18f05dd8eb50e](https://github.com/cockroachdb/cockroach/commits/cc6ca026319024800395293b0fb18f05dd8eb50e):


```
=== RUN   TestStoreRangeSplitAndMergeWithGlobalReads
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/3c632028e5b231a582105881dfd5d4b5/logTestStoreRangeSplitAndMergeWithGlobalReads54447473
    test_log_scope.go:81: use -show-logs to present logs inline
    client_split_test.go:3691: condition failed to evaluate within 45s: from client_split_test.go:3693: commit wait count is 2
    panic.go:523: -- test log scope end --
--- FAIL: TestStoreRangeSplitAndMergeWithGlobalReads (53.05s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeSplitAndMergeWithGlobalReads.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-36100

kv/kvserver.TestStoreRangeSplitAndMergeWithGlobalReads [failed](https://tanzanite.cluster.engflow.com/invocations/default/e0d64c03-bbe8-4af3-b0cb-5866cef06f13?testReportRun=9&testReportShard=11&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [617bf347978dcc0d711399b1a76402d7f88de958](https://github.com/cockroachdb/cockroach/commits/617bf347978dcc0d711399b1a76402d7f88de958):


```
=== RUN   TestStoreRangeSplitAndMergeWithGlobalReads
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestStoreRangeSplitAndMergeWithGlobalReads331702955
    test_log_scope.go:81: use -show-logs to present logs inline
    client_split_test.go:3691: condition failed to evaluate within 45s: from client_split_test.go:3693: commit wait count is 2
    panic.go:523: -- test log scope end --
test logs left over in: outputs.zip/logTestStoreRangeSplitAndMergeWithGlobalReads331702955
--- FAIL: TestStoreRangeSplitAndMergeWithGlobalReads (52.99s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=9</code>
 - <code>shard=11</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeSplitAndMergeWithGlobalReads.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestStoreRangeSplitAndMergeWithGlobalReads [failed](https://tanzanite.cluster.engflow.com/invocations/default/01563976-1fe2-4191-bba5-22eea9550bd0?testReportRun=12&testReportShard=11&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [0b7ae19e2b94b851ed8812914f57032aab699811](https://github.com/cockroachdb/cockroach/commits/0b7ae19e2b94b851ed8812914f57032aab699811):


```
=== RUN   TestStoreRangeSplitAndMergeWithGlobalReads
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestStoreRangeSplitAndMergeWithGlobalReads2767376744
    test_log_scope.go:81: use -show-logs to present logs inline
    client_split_test.go:3691: condition failed to evaluate within 45s: from client_split_test.go:3693: commit wait count is 2
    panic.go:523: -- test log scope end --
test logs left over in: outputs.zip/logTestStoreRangeSplitAndMergeWithGlobalReads2767376744
--- FAIL: TestStoreRangeSplitAndMergeWithGlobalReads (53.01s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=12</code>
 - <code>shard=11</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeSplitAndMergeWithGlobalReads.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestStoreRangeSplitAndMergeWithGlobalReads [failed](https://tanzanite.cluster.engflow.com/invocations/default/5d1a7a35-61ed-4420-9f70-061b293c881f?testReportRun=23&testReportShard=11&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [a36097be277adef635f55d317579ca79b450bfef](https://github.com/cockroachdb/cockroach/commits/a36097be277adef635f55d317579ca79b450bfef):


```
=== RUN   TestStoreRangeSplitAndMergeWithGlobalReads
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestStoreRangeSplitAndMergeWithGlobalReads3287400935
    test_log_scope.go:81: use -show-logs to present logs inline
    client_split_test.go:3691: condition failed to evaluate within 45s: from client_split_test.go:3693: commit wait count is 2
    panic.go:523: -- test log scope end --
test logs left over in: outputs.zip/logTestStoreRangeSplitAndMergeWithGlobalReads3287400935
--- FAIL: TestStoreRangeSplitAndMergeWithGlobalReads (53.01s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=23</code>
 - <code>shard=11</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeSplitAndMergeWithGlobalReads.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestStoreRangeSplitAndMergeWithGlobalReads [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/14074179?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/14074179?buildTab=artifacts#/) on master @ [a78e1972a82f5b1bbb50d715aff46f7b668036fe](https://github.com/cockroachdb/cockroach/commits/a78e1972a82f5b1bbb50d715aff46f7b668036fe):


```
=== RUN   TestStoreRangeSplitAndMergeWithGlobalReads
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/3c632028e5b231a582105881dfd5d4b5/logTestStoreRangeSplitAndMergeWithGlobalReads549076652
    test_log_scope.go:81: use -show-logs to present logs inline
    client_split_test.go:3691: condition failed to evaluate within 45s: from client_split_test.go:3693: commit wait count is 2
    panic.go:523: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/3c632028e5b231a582105881dfd5d4b5/logTestStoreRangeSplitAndMergeWithGlobalReads549076652
--- FAIL: TestStoreRangeSplitAndMergeWithGlobalReads (53.06s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeSplitAndMergeWithGlobalReads.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestStoreRangeSplitAndMergeWithGlobalReads [failed](https://tanzanite.cluster.engflow.com/invocations/default/7f1d73c5-5ddd-4855-bf60-ae0b8b268601?testReportRun=15&testReportShard=11&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [e50b0ec4d3a53f81e26f3776ae3f3be55d435a9a](https://github.com/cockroachdb/cockroach/commits/e50b0ec4d3a53f81e26f3776ae3f3be55d435a9a):


```
=== RUN   TestStoreRangeSplitAndMergeWithGlobalReads
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestStoreRangeSplitAndMergeWithGlobalReads4231620982
    test_log_scope.go:81: use -show-logs to present logs inline
    client_split_test.go:3691: condition failed to evaluate within 45s: from client_split_test.go:3693: commit wait count is 2
    panic.go:523: -- test log scope end --
test logs left over in: outputs.zip/logTestStoreRangeSplitAndMergeWithGlobalReads4231620982
--- FAIL: TestStoreRangeSplitAndMergeWithGlobalReads (53.00s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=15</code>
 - <code>shard=11</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeSplitAndMergeWithGlobalReads.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestStoreRangeSplitAndMergeWithGlobalReads [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/14094510?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/14094510?buildTab=artifacts#/) on master @ [31acb7a07a4e6e1e96ceb8533cfa042ea80514a8](https://github.com/cockroachdb/cockroach/commits/31acb7a07a4e6e1e96ceb8533cfa042ea80514a8):


```
=== RUN   TestStoreRangeSplitAndMergeWithGlobalReads
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/93fbc720f61b7e85f714b9750f754229/logTestStoreRangeSplitAndMergeWithGlobalReads180609409
    test_log_scope.go:81: use -show-logs to present logs inline
    client_split_test.go:3691: condition failed to evaluate within 45s: from client_split_test.go:3693: commit wait count is 2
    panic.go:523: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/93fbc720f61b7e85f714b9750f754229/logTestStoreRangeSplitAndMergeWithGlobalReads180609409
--- FAIL: TestStoreRangeSplitAndMergeWithGlobalReads (53.01s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeSplitAndMergeWithGlobalReads.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestStoreRangeSplitAndMergeWithGlobalReads [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/14106901?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/14106901?buildTab=artifacts#/) on master @ [ee3168ac3e0286a63dd49ab8b9f14b036ad23bde](https://github.com/cockroachdb/cockroach/commits/ee3168ac3e0286a63dd49ab8b9f14b036ad23bde):


```
=== RUN   TestStoreRangeSplitAndMergeWithGlobalReads
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/93fbc720f61b7e85f714b9750f754229/logTestStoreRangeSplitAndMergeWithGlobalReads1817694607
    test_log_scope.go:81: use -show-logs to present logs inline
    client_split_test.go:3691: condition failed to evaluate within 45s: from client_split_test.go:3693: commit wait count is 2
    panic.go:523: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/93fbc720f61b7e85f714b9750f754229/logTestStoreRangeSplitAndMergeWithGlobalReads1817694607
--- FAIL: TestStoreRangeSplitAndMergeWithGlobalReads (53.37s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeSplitAndMergeWithGlobalReads.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestStoreRangeSplitAndMergeWithGlobalReads [failed](https://tanzanite.cluster.engflow.com/invocations/default/90411a40-d13d-44c1-bee6-2218265e598c?testReportRun=24&testReportShard=12&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [6d65201b9b603e0b3fcf1d509ec23edfdd68de45](https://github.com/cockroachdb/cockroach/commits/6d65201b9b603e0b3fcf1d509ec23edfdd68de45):


```
=== RUN   TestStoreRangeSplitAndMergeWithGlobalReads
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestStoreRangeSplitAndMergeWithGlobalReads3385134434
    test_log_scope.go:81: use -show-logs to present logs inline
    client_split_test.go:3691: condition failed to evaluate within 45s: from client_split_test.go:3693: commit wait count is 2
    panic.go:523: -- test log scope end --
test logs left over in: outputs.zip/logTestStoreRangeSplitAndMergeWithGlobalReads3385134434
--- FAIL: TestStoreRangeSplitAndMergeWithGlobalReads (52.98s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=24</code>
 - <code>shard=12</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeSplitAndMergeWithGlobalReads.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/BUILD.bazel
+++ b/pkg/kv/kvserver/BUILD.bazel
@@ -454,6 +454,7 @@ go_test(
         "//pkg/spanconfig",
         "//pkg/spanconfig/spanconfigptsreader",
         "//pkg/spanconfig/spanconfigstore",
+        "//pkg/spanconfig/spanconfigtestutils",
         "//pkg/sql",
         "//pkg/sql/catalog/bootstrap",
         "//pkg/sql/catalog/catalogkeys",

--- a/pkg/kv/kvserver/client_spanconfigs_test.go
+++ b/pkg/kv/kvserver/client_spanconfigs_test.go
@@ -55,7 +55,7 @@ func TestSpanConfigUpdateAppliedToReplica(t *testing.T) {
 				DisableGCQueue:    true,
 			},
 			SpanConfig: &spanconfig.TestingKnobs{
-				StoreKVSubscriberOverride: mockSubscriber,
+				StoreKVSubscriberOverride: func(spanconfig.KVSubscriber) spanconfig.KVSubscriber { return mockSubscriber },
 			},
 		},
 	}
@@ -120,7 +120,7 @@ func TestFallbackSpanConfigOverride(t *testing.T) {
 				DisableGCQueue:    true,
 			},
 			SpanConfig: &spanconfig.TestingKnobs{
-				StoreKVSubscriberOverride: mockSubscriber,
+				StoreKVSubscriberOverride: func(spanconfig.KVSubscriber) spanconfig.KVSubscriber { return mockSubscriber },
 			},
 		},
 	}

--- a/pkg/kv/kvserver/client_split_test.go
+++ b/pkg/kv/kvserver/client_split_test.go
@@ -4051,7 +4051,8 @@ func TestStoreRangeSplitAndMergeWithGlobalReads(t *testing.T) {
 	spanConfig.GlobalReads = true
 
 	ctx := context.Background()
-	s, sqlDB, _ := serverutils.StartServer(t, base.TestServerArgs{
+	s := serverutils.StartServerOnly(t, base.TestServerArgs{
+		DisableSQLServer: true,
 		Knobs: base.TestingKnobs{
 			Store: &kvserver.StoreTestingKnobs{
 				DisableMergeQueue:     true,
@@ -4071,10 +4072,6 @@ func TestStoreRangeSplitAndMergeWithGlobalReads(t *testing.T) {
 	defer s.Stopper().Stop(ctx)
 	// Set the closed_timestamp interval to be short to shorten the test duration
 	// because we need to wait for a checkpoint on the system config.
-	tdb := sqlutils.MakeSQLRunner(sqlDB)
-	tdb.Exec(t, `SET CLUSTER SETTING kv.closed_timestamp.target_duration = '20ms'`)
-	tdb.Exec(t, `SET CLUSTER SETTING kv.closed_timestamp.side_transport_interval = '20ms'`)
-	tdb.Exec(t, `SET CLUSTER SETTING kv.rangefeed.closed_timestamp_refresh_interval = '20ms'`)
 	store, err := s.GetStores().(*kvserver.Stores).GetStore(s.GetFirstStoreID())
 	require.NoError(t, err)
 

--- a/pkg/kv/kvserver/client_split_test.go
+++ b/pkg/kv/kvserver/client_split_test.go
@@ -4042,17 +4042,10 @@ func TestStoreRangeSplitAndMergeWithGlobalReads(t *testing.T) {
 		}
 		return nil
 	}
-	// Set global reads.
-	zoneConfig := zonepb.DefaultZoneConfig()
-	zoneConfig.GlobalReads = proto.Bool(true)
 
 	ctx := context.Background()
 	s, sqlDB, _ := serverutils.StartServer(t, base.TestServerArgs{
 		Knobs: base.TestingKnobs{
-			Server: &server.TestingKnobs{
-				DefaultZoneConfigOverride: &zoneConfig,
-			},
-
 			Store: &kvserver.StoreTestingKnobs{
 				DisableMergeQueue:     true,
 				TestingResponseFilter: respFilter,
@@ -4082,6 +4075,11 @@ func TestStoreRangeSplitAndMergeWithGlobalReads(t *testing.T) {
 	// response filter.
 	clockPtr.Store(s.Clock())
 
+	// Set global reads.
+	zoneConfig := zonepb.DefaultZoneConfig()
+	zoneConfig.GlobalReads = proto.Bool(true)
+	config.TestingSetZoneConfig(config.ObjectID(descID), zoneConfig)
+
 	// Perform a write to the system config span being watched by
 	// the SystemConfigProvider.
 	tdb.Exec(t, "CREATE TABLE foo ()")
@@ -4089,19 +4087,6 @@ func TestStoreRangeSplitAndMergeWithGlobalReads(t *testing.T) {
 		repl := store.LookupReplica(roachpb.RKey(descKey))
 		if repl.ClosedTimestampPolicy() != roachpb.LEAD_FOR_GLOBAL_READS {
 			return errors.Errorf("expected LEAD_FOR_GLOBAL_READS policy")
-		}
-		return nil
-	})
-
-	// The commit wait count is 1 due to the split above since global reads are
-	// set for the default config.
-	var splitCount = int64(1)
-	testutils.SucceedsSoon(t, func() error {
-		if splitCount != store.Metrics().CommitWaitsBeforeCommitTrigger.Count() {
-			return errors.Errorf("commit wait count is %d", store.Metrics().CommitWaitsBeforeCommitTrigger.Count())
-		}
-		if splitCount != atomic.LoadInt64(&splits) {
-			return errors.Errorf("num splits is %d", atomic.LoadInt64(&splits))
 		}
 		return nil
 	})
@@ -4116,9 +4101,8 @@ func TestStoreRangeSplitAndMergeWithGlobalReads(t *testing.T) {
 	splitArgs = adminSplitArgs(splitKey)
 	_, pErr = kv.SendWrapped(ctx, store.TestSender(), splitArgs)
 	require.Nil(t, pErr)
-	splitCount++
-	require.Equal(t, splitCount, store.Metrics().CommitWaitsBeforeCommitTrigger.Count())
-	require.Equal(t, splitCount, atomic.LoadInt64(&splits))
+	require.Equal(t, int64(1), store.Metrics().CommitWaitsBeforeCommitTrigger.Count())
+	require.Equal(t, int64(1), atomic.LoadInt64(&splits))
 
 	repl := store.LookupReplica(roachpb.RKey(splitKey))
 	require.Equal(t, splitKey, repl.Desc().StartKey.AsRawKey())
@@ -4127,7 +4111,7 @@ func TestStoreRangeSplitAndMergeWithGlobalReads(t *testing.T) {
 	mergeArgs := adminMergeArgs(descKey)
 	_, pErr = kv.SendWrapped(ctx, store.TestSender(), mergeArgs)
 	require.Nil(t, pErr)
-	require.Equal(t, splitCount+1, store.Metrics().CommitWaitsBeforeCommitTrigger.Count())
+	require.Equal(t, int64(2), store.Metrics().CommitWaitsBeforeCommitTrigger.Count())
 	require.Equal(t, int64(1), atomic.LoadInt64(&merges))
 
 	repl = store.LookupReplica(roachpb.RKey(splitKey))

--- a/pkg/kv/kvserver/client_split_test.go
+++ b/pkg/kv/kvserver/client_split_test.go
@@ -51,6 +51,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/server"
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
 	"github.com/cockroachdb/cockroach/pkg/spanconfig"
+	"github.com/cockroachdb/cockroach/pkg/spanconfig/spanconfigtestutils"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/bootstrap"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/systemschema"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlstats"
@@ -4007,8 +4008,6 @@ func TestStoreRangeSplitAndMergeWithGlobalReads(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
-	skip.WithIssue(t, 119230)
-
 	// Detect splits and merges over the global read ranges. Assert that the split
 	// and merge transactions commit with pushed write timestamps, and that the
 	// commit-wait sleep for these transactions is performed before running their
@@ -4043,12 +4042,28 @@ func TestStoreRangeSplitAndMergeWithGlobalReads(t *testing.T) {
 		return nil
 	}
 
+	descID := bootstrap.TestingUserDescID(0)
+	descKey := keys.SystemSQLCodec.TablePrefix(descID)
+	splitKey := append(descKey, []byte("split")...)
+
+	// Set global reads for the test ranges.
+	spanConfig := zonepb.DefaultZoneConfig().AsSpanConfig()
+	spanConfig.GlobalReads = true
+
 	ctx := context.Background()
 	s, sqlDB, _ := serverutils.StartServer(t, base.TestServerArgs{
 		Knobs: base.TestingKnobs{
 			Store: &kvserver.StoreTestingKnobs{
 				DisableMergeQueue:     true,
 				TestingResponseFilter: respFilter,
+			},
+			SpanConfig: &spanconfig.TestingKnobs{
+				StoreKVSubscriberOverride: func(original spanconfig.KVSubscriber) spanconfig.KVSubscriber {
+					wrapped := spanconfigtestutils.NewWrappedKVSubscriber(original)
+					wrapped.AddOverride(descKey, spanConfig)
+					wrapped.AddOverride(splitKey, spanConfig)
+					return wrapped
+				},
 			},
 		},
 	})
@@ -4062,11 +4077,8 @@ func TestStoreRangeSplitAndMergeWithGlobalReads(t *testing.T) {
 	tdb.Exec(t, `SET CLUSTER SETTING kv.rangefeed.closed_timestamp_refresh_interval = '20ms'`)
 	store, err := s.GetStores().(*kvserver.Stores).GetStore(s.GetFirstStoreID())
 	require.NoError(t, err)
-	config.TestingSetupZoneConfigHook(s.Stopper())
 
 	// Split off the range for the test.
-	descID := bootstrap.TestingUserDescID(0)
-	descKey := keys.SystemSQLCodec.TablePrefix(descID)
 	splitArgs := adminSplitArgs(descKey)
 	_, pErr := kv.SendWrapped(ctx, store.TestSender(), splitArgs)
 	require.Nil(t, pErr)
@@ -4075,21 +4087,9 @@ func TestStoreRangeSplitAndMergeWithGlobalReads(t *testing.T) {
 	// response filter.
 	clockPtr.Store(s.Clock())
 
-	// Set global reads.
-	zoneConfig := zonepb.DefaultZoneConfig()
-	zoneConfig.GlobalReads = proto.Bool(true)
-	config.TestingSetZoneConfig(config.ObjectID(descID), zoneConfig)
-
-	// Perform a write to the system config span being watched by
-	// the SystemConfigProvider.
-	tdb.Exec(t, "CREATE TABLE foo ()")
-	testutils.SucceedsSoon(t, func() error {
-		repl := store.LookupReplica(roachpb.RKey(descKey))
-		if repl.ClosedTimestampPolicy() != roachpb.LEAD_FOR_GLOBAL_READS {
-			return errors.Errorf("expected LEAD_FOR_GLOBAL_READS policy")
-		}
-		return nil
-	})
+	// Verify that the closed timestamp policy is set up.
+	repl := store.LookupReplica(roachpb.RKey(descKey))
+	require.Equal(t, repl.ClosedTimestampPolicy(), roachpb.LEAD_FOR_GLOBAL_READS)
 
 	// Write to the range, which has the effect of bumping the closed timestamp.
 	pArgs := putArgs(descKey, []byte("foo"))
@@ -4097,14 +4097,13 @@ func TestStoreRangeSplitAndMergeWithGlobalReads(t *testing.T) {
 	require.Nil(t, pErr)
 
 	// Split the range. Should succeed.
-	splitKey := append(descKey, []byte("split")...)
 	splitArgs = adminSplitArgs(splitKey)
 	_, pErr = kv.SendWrapped(ctx, store.TestSender(), splitArgs)
 	require.Nil(t, pErr)
 	require.Equal(t, int64(1), store.Metrics().CommitWaitsBeforeCommitTrigger.Count())
 	require.Equal(t, int64(1), atomic.LoadInt64(&splits))
 
-	repl := store.LookupReplica(roachpb.RKey(splitKey))
+	repl = store.LookupReplica(roachpb.RKey(splitKey))
 	require.Equal(t, splitKey, repl.Desc().StartKey.AsRawKey())
 
 	// Merge the range. Should succeed.

--- a/pkg/server/server.go
+++ b/pkg/server/server.go
@@ -753,41 +753,41 @@ func NewServer(cfg Config, stopper *stop.Stopper) (serverctl.ServerStartupInterf
 		kvAccessorForTenantRecords spanconfig.KVAccessor
 	}
 	spanConfigKnobs, _ := cfg.TestingKnobs.SpanConfig.(*spanconfig.TestingKnobs)
-	if spanConfigKnobs != nil && spanConfigKnobs.StoreKVSubscriberOverride != nil {
-		spanConfig.subscriber = spanConfigKnobs.StoreKVSubscriberOverride
-	} else {
-		// We use the span configs infra to control whether rangefeeds are
-		// enabled on a given range. At the moment this only applies to
-		// system tables (on both host and secondary tenants). We need to
-		// consider two things:
-		// - The sql-side reconciliation process runs asynchronously. When
-		//   the config for a given range is requested, we might not yet have
-		//   it, thus falling back to the static config below.
-		// - Various internal subsystems rely on rangefeeds to function.
-		//
-		// Consequently, we configure our static fallback config to actually
-		// allow rangefeeds. As the sql-side reconciliation process kicks
-		// off, it'll install the actual configs that we'll later consult.
-		// For system table ranges we install configs that allow for
-		// rangefeeds. Until then, we simply allow rangefeeds when a more
-		// targeted config is not found.
-		fallbackConf := cfg.DefaultZoneConfig.AsSpanConfig()
-		fallbackConf.RangefeedEnabled = true
-		// We do the same for opting out of strict GC enforcement; it
-		// really only applies to user table ranges
-		fallbackConf.GCPolicy.IgnoreStrictEnforcement = true
+	// We use the span configs infra to control whether rangefeeds are
+	// enabled on a given range. At the moment this only applies to
+	// system tables (on both host and secondary tenants). We need to
+	// consider two things:
+	// - The sql-side reconciliation process runs asynchronously. When
+	//   the config for a given range is requested, we might not yet have
+	//   it, thus falling back to the static config below.
+	// - Various internal subsystems rely on rangefeeds to function.
+	//
+	// Consequently, we configure our static fallback config to actually
+	// allow rangefeeds. As the sql-side reconciliation process kicks
+	// off, it'll install the actual configs that we'll later consult.
+	// For system table ranges we install configs that allow for
+	// rangefeeds. Until then, we simply allow rangefeeds when a more
+	// targeted config is not found.
+	fallbackConf := cfg.DefaultZoneConfig.AsSpanConfig()
+	fallbackConf.RangefeedEnabled = true
+	// We do the same for opting out of strict GC enforcement; it
+	// really only applies to user table ranges
+	fallbackConf.GCPolicy.IgnoreStrictEnforcement = true
 
-		spanConfig.subscriber = spanconfigkvsubscriber.New(
-			clock,
-			rangeFeedFactory,
-			keys.SpanConfigurationsTableID,
-			4<<20, /* 4 MB */
-			fallbackConf,
-			cfg.Settings,
-			spanconfigstore.NewBoundsReader(tenantCapabilitiesWatcher),
-			spanConfigKnobs,
-			nodeRegistry,
-		)
+	spanConfig.subscriber = spanconfigkvsubscriber.New(
+		clock,
+		rangeFeedFactory,
+		keys.SpanConfigurationsTableID,
+		4<<20, /* 4 MB */
+		fallbackConf,
+		cfg.Settings,
+		spanconfigstore.NewBoundsReader(tenantCapabilitiesWatcher),
+		spanConfigKnobs,
+		nodeRegistry,
+	)
+
+	if spanConfigKnobs != nil && spanConfigKnobs.StoreKVSubscriberOverride != nil {
+		spanConfig.subscriber = spanConfigKnobs.StoreKVSubscriberOverride(spanConfig.subscriber)
 	}
 
 	scKVAccessor := spanconfigkvaccessor.New(

--- a/pkg/spanconfig/spanconfigtestutils/BUILD.bazel
+++ b/pkg/spanconfig/spanconfigtestutils/BUILD.bazel
@@ -21,6 +21,7 @@ go_library(
         "//pkg/sql/isql",
         "//pkg/util/hlc",
         "//pkg/util/syncutil",
+        "//pkg/util/timeutil",
         "@com_github_cockroachdb_datadriven//:datadriven",
         "@com_github_stretchr_testify//require",
         "@in_gopkg_yaml_v2//:yaml_v2",

--- a/pkg/spanconfig/spanconfigtestutils/utils.go
+++ b/pkg/spanconfig/spanconfigtestutils/utils.go
@@ -11,6 +11,7 @@
 package spanconfigtestutils
 
 import (
+	"context"
 	"fmt"
 	"reflect"
 	"regexp"
@@ -28,6 +29,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/spanconfig/spanconfigbounds"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/datadriven"
 	"github.com/stretchr/testify/require"
 	"gopkg.in/yaml.v2"
@@ -750,3 +752,82 @@ func ParseProtectionTarget(t testing.TB, input string) *ptpb.Target {
 	}
 	return nil
 }
+
+// configOverride is used to override span configs for specific ranges.
+type configOverride struct {
+	key    roachpb.Key
+	config roachpb.SpanConfig
+}
+
+// WrappedKVSubscriber is a KVSubscriber which wraps another KVSubscriber and
+// overrides specific SpanConfigs. This test struct should be used in
+// StoreKVSubscriberOverride as a more generic and powerful SpanConfig injection
+// alternative to change the behavior for specific ranges instead of using any
+// of DefaultZoneConfigOverride, DefaultSystemZoneConfigOverride,
+// OverrideFallbackConf, ConfReaderInterceptor, UseSystemConfigSpanForQueues,
+// SpanConfigUpdateInterceptor or SetSpanConfigInterceptor. By using this struct
+// as an alternative to the others it mainly avoids any need to depend on
+// rangefeed timing to propagate the config change.
+type WrappedKVSubscriber struct {
+	wrapped spanconfig.KVSubscriber
+	// Overrides are list of tuples of roachpb.Key and spanconfig.
+	overrides []configOverride
+}
+
+// NewWrappedKVSubscriber creates a new WrappedKVSubscriber.
+func NewWrappedKVSubscriber(wrapped spanconfig.KVSubscriber) *WrappedKVSubscriber {
+	return &WrappedKVSubscriber{wrapped: wrapped}
+}
+
+// AddOverride adds a new override to the WrappedKVSubscriber. This should only
+// be used during construction.
+func (w *WrappedKVSubscriber) AddOverride(key roachpb.Key, config roachpb.SpanConfig) {
+	w.overrides = append(w.overrides, configOverride{key: key, config: config})
+}
+
+// GetProtectionTimestamps implements spanconfig.KVSubscriber.
+func (w *WrappedKVSubscriber) GetProtectionTimestamps(
+	ctx context.Context, sp roachpb.Span,
+) ([]hlc.Timestamp, hlc.Timestamp, error) {
+	return w.wrapped.GetProtectionTimestamps(ctx, sp)
+}
+
+// LastUpdated always reports that it has been updated to allow
+// GetSpanConfigForKey to be used immediately.
+func (w *WrappedKVSubscriber) LastUpdated() hlc.Timestamp {
+	return hlc.Timestamp{WallTime: timeutil.Now().UnixNano()}
+}
+
+// Subscribe implements spanconfig.KVSubscriber.
+func (w *WrappedKVSubscriber) Subscribe(f func(context.Context, roachpb.Span)) {
+	w.wrapped.Subscribe(f)
+}
+
+// ComputeSplitKey implements spanconfig.StoreReader.
+func (w *WrappedKVSubscriber) ComputeSplitKey(
+	ctx context.Context, start roachpb.RKey, end roachpb.RKey,
+) (roachpb.RKey, error) {
+	return w.wrapped.ComputeSplitKey(ctx, start, end)
+}
+
+// GetSpanConfigForKey implements spanconfig.StoreReader.
+func (w *WrappedKVSubscriber) GetSpanConfigForKey(
+	ctx context.Context, key roachpb.RKey,
+) (roachpb.SpanConfig, roachpb.Span, error) {
+	spanConfig, span, err := w.wrapped.GetSpanConfigForKey(ctx, key)
+	for _, o := range w.overrides {
+		if key.Equal(o.key) {
+			return o.config, span, nil
+		}
+	}
+	return spanConfig, span, err
+}
+
+// NeedsSplit implements spanconfig.StoreReader.
+func (w *WrappedKVSubscriber) NeedsSplit(
+	ctx context.Context, start roachpb.RKey, end roachpb.RKey,
+) (bool, error) {
+	return w.wrapped.NeedsSplit(ctx, start, end)
+}
+
+var _ spanconfig.KVSubscriber = &WrappedKVSubscriber{}

--- a/pkg/spanconfig/testing_knobs.go
+++ b/pkg/spanconfig/testing_knobs.go
@@ -62,7 +62,7 @@ type TestingKnobs struct {
 
 	// StoreKVSubscriberOverride is used to override the KVSubscriber used when
 	// setting up a new store.
-	StoreKVSubscriberOverride KVSubscriber
+	StoreKVSubscriberOverride func(KVSubscriber) KVSubscriber
 
 	// KVAccessorPaginationInterceptor, if set, is invoked on every pagination
 	// event.

